### PR TITLE
Make pc private in goto_trace_stept

### PR DESCRIPTION
**Pull request by reuk**
See diffblue/test-gen#983 for an explanation.

### CODE REVIEW

File: src/goto-programs/goto_trace.h [hunk 1]
```diff
@@ -142,6 +140,22 @@ class goto_trace_stept
...
     cond_expr.make_nil();
   }
+
+  goto_programt::const_targett get_pc() const { return pc; }
+
+  /// We don't have a way of enforcing this, but \p pc_ must *not* be a
+  /// singular iterator!
+  void set_pc(goto_programt::const_targett pc_)
+  {
+    pc = std::move(pc_);
```
Comment (thk123) [inline src/goto-programs/goto_trace.h]: Nit: whitespace

File: src/goto-programs/goto_program.h [hunk 2]
```diff
@@ -82,6 +82,13 @@ inline bool operator<(
   return order_const_target<codet, exprt>(i1, i2);
 }
 
+inline bool operator<(
```
Comment (thk123) [inline src/goto-programs/goto_program.h]: This change seems apropos of nothing, how come it is introduced in this PR?

File: src/goto-programs/goto_trace.h (hunk 1 above)
Comment (reuk) [inline src/goto-programs/goto_trace.h]: Well spotted! Will fix

File: src/goto-programs/goto_program.h (hunk 2 above)
Comment (reuk) [inline src/goto-programs/goto_program.h]: This duplicates https://github.com/diffblue/cbmc/pull/1247, but I needed it in order to build - once 1247 is merged I'll remove this

**Comment by cesaro** [comment]
We already discussed this change orally. Think about my suggestion. Another reason in favor of my proposal is that it addresses the core of the problem, which is initialization, rather than encapsulation:

- Encapsulating `pc` and offering _the_ setter and getter you provide doesn't fix the underlying issue (would not say this if your getter checks the Boolean first, but that also has a (minor) performance penalty), as the user of the trace step needs to still be aware of when to test for the Boolean. This solution increases encapsulation, increases mechanism (getter vs direct reading), but does not augment decoupling.

**Comment by reuk** [comment]
A problem with comparing against an iterator of a global static collection is that iterator comparisons between iterators of different collections is UB

**Comment by reuk** [comment]
I've fixed the formatting issues and removed the change from #1247. Could one of you hit merge for me please?

**Pull request merged**